File: src/Core/Managers/Tasks/TaskManager.java
package Core.Managers.Tasks;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

/**
 * Created by steppers on 28/10/15.
 */
public class TaskManager {

    private BlockingQueue<Task> taskQueue = new LinkedBlockingQueue<>();
    private TaskThread threads[];

    public TaskManager()
    {
        threads = new TaskThread[Runtime.getRuntime().availableProcessors()];
        for(int i = 0; i < threads.length; i++)
        {
            threads[i] = new TaskThread();
            threads[i].start();
        }
    }

    public void submitTask(Task task)
    {
        try {
            taskQueue.put(task);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void dispose()
    {
        for(TaskThread thread : threads)
        {
            thread.closeThread();
        }
    }

    private class TaskThread extends Thread
    {
        public Task t = null;

        public void run()
        {
            while(true) {
                try {
                    t = taskQueue.take();
                    t.execute();
                } catch (InterruptedException e) {
                    break;
                }
            }
        }

        public synchronized void closeThread()
        {
            interrupt();
        }
    }
}

File: CPP-Version/src/Engine/Managers/TaskManager/TaskManager.cpp
//
// Created by steppers on 05/11/15.
//

#include "TaskManager.h"
#include "TerminateTask.h"

TaskManager::TaskManager() { }
TaskManager::~TaskManager() { }

void TaskManager::init() {
    int numCores = thread::hardware_concurrency();
    for(int i = 0; i < numCores; i++)
    {
        _threads.push_back(new TaskThread(i, this));
    }
}

void TaskManager::submitTask(Task *task) {
    _tasks.add(task);
}

void TaskManager::submitTasks(vector<Task*> tasks) {
    for(Task* t : tasks)
        _tasks.add(t);
}

void TaskManager::killThreads() {
    for(TaskThread* t : _threads)
        _tasks.add(new TerminateTask());
    for(TaskThread* t : _threads)
        t->join();
}

void TaskManager::waitForTasks() {

}

Task* TaskManager::takeTask() {
    return _tasks.take();
}